File: day23_part2_2020.dart

import 'dart:io';

const totalCups = 1000000;
const totalMoves = 10000000;

void main() {
  var file = new File('input.txt');
  var input = file.readAsStringSync().trim();

  var cups = List<int>.filled(totalCups + 1, 0);
  var lastCup = 0;

  for (var i = 0; i < input.length; i++) {
    var cup = int.parse(input[i]);
    if (i > 0) {
      cups[lastCup] = cup;
    }
    lastCup = cup;
  }

  for (var i = input.length + 1; i <= totalCups; i++) {
    cups[lastCup] = i;
    lastCup = i;
  }
  cups[lastCup] = int.parse(input[0]);

  var currentCup = int.parse(input[0]);
  for (var i = 0; i < totalMoves; i++) {
    var pickup1 = cups[currentCup];
    var pickup2 = cups[pickup1];
    var pickup3 = cups[pickup2];

    cups[currentCup] = cups[pickup3];

    var destinationCup = currentCup - 1;
    if (destinationCup == 0) {
      destinationCup = totalCups;
    }
    while (destinationCup == pickup1 || destinationCup == pickup2 || destinationCup == pickup3) {
      destinationCup--;
      if (destinationCup == 0) {
        destinationCup = totalCups;
      }
    }

    cups[pickup3] = cups[destinationCup];
    cups[destinationCup] = pickup1;

    currentCup = cups[currentCup];
  }

  var cup1 = cups[1];
  var cup2 = cups[cup1];
  print(cup1 * cup2);
}

File: training_data/reflection-tuning/day23_part2_2020.rb
def simulate_crab_cups(input, num_cups, num_moves)
  # Initialize cups array
  cups = Array.new(num_cups + 1, 0)
  first_cup = input[0].to_i
  input.each_char.with_index do |cup, i|
    next_cup = i == input.length - 1 ? first_cup : input[i + 1].to_i
    cups[cup.to_i] = next_cup
  end

  if num_cups > input.length
    last_input_cup = input[-1].to_i
    cups[last_input_cup] = input.length + 1
    (input.length + 1..num_cups).each { |i| cups[i] = i + 1 }
    cups[num_cups] = first_cup
  end

  current = first_cup

  num_moves.times do
    # Pick up three cups
    pickup1 = cups[current]
    pickup2 = cups[pickup1]
    pickup3 = cups[pickup2]
    next_current = cups[pickup3]

    # Find destination
    destination = current - 1
    destination = num_cups if destination == 0
    while [pickup1, pickup2, pickup3].include?(destination)
      destination -= 1
      destination = num_cups if destination == 0
    end

    # Place picked up cups after destination
    cups[current] = next_current
    cups[pickup3] = cups[destination]
    cups[destination] = pickup1

    current = next_current
  end

  cups
end

# Read input from file
input = File.read('input.txt').strip

# Part 1
cups = simulate_crab_cups(input, input.length, 100)
result = []
current = cups[1]
8.times do
  result << current
  current = cups[current]
end
puts "Part 1: #{result.join}"

# Part 2
cups = simulate_crab_cups(input, 1_000_000, 10_000_000)
cup1 = cups[1]
cup2 = cups[cup1]
puts "Part 2: #{cup1 * cup2}"
